Add GPU implementation for Unique[WithCounts] ops

benbarsdell opened the pull request:
Adds a GPU op kernel supporting the Unique and UniqueWithCounts ops (not the V2 versions).

cc @nluehr

benbarsdell committed:
Enable GPU radix sort with half and bfloat16 types

- This enables it for all users of gpu_prim.h instead of just in the
  TopK op implementation.

benbarsdell committed:
Add convenience wrappers for GPU sort and scan

- This will avoid code duplication in kernels that need them.

benbarsdell committed:
Add GPU implementation for Unique[WithCounts]

benbarsdell committed:
Add empty-input test to unique_op_test.py

- This ensures the op works when operating on a zero-sized input tensor.

benbarsdell commented:
Changed to WIP because I see from the test failures that I made an oversight in the contract of this op (I only ran the main tests myself and they didn't catch it). I will update again once I have a fix for it.

benbarsdell committed:
Add additional tests for Unique[WithCounts]

- The existing tests in this file don't check the exact output,
  which must have values in the order of first appearance.

benbarsdell committed:
Fix Unique GPU implem to order by first appearance

- The previous implementation ordered the outputs by value rather than
  by order of first appearance as required for this op.
- Also applies to UniqueWithCounts.

benbarsdell commented:
I've changed the implementation to produce the output in the correct order now. (I also added an explicit test for that).

Removing "WIP", ready for review.

sanjoy commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,443 @@
+
+    int64 input_size = input.NumElements();
+
+    // The algorithm implemented here is as follows:
+    // 1) Sort the input to group equal values together in segments.
+    //      sorted_input, sorted_input_inds = sort(input)
+    // 2) Identify the boundaries between segments and use prefix sum to
+    //    compute the unique ID for each sorted value.
+    //      sorted_input_unique_ids = prefix_sum(indicator(sorted_input))
+    // 3) Extract the input index of the first occurrence of each unique value.
+    //    If counts are required, also extract the end index of each segment.
+    //      unique_input_inds[sorted_input_unique_ids] =
+    //          sorted_input_inds (@ indicator)
+    //      segment_ends[sorted_input_unique_ids[i] - 1] = i (@ indicator)
+    // 4) Sort the extracted unique input indices to put them in order of
+    //    first appearance.
+    //      sorted_unique_input_inds, sorted_unique_perm =
+    //          sort(unique_input_inds)
+    // 5) Gather the sorted unique input values to produce output, and invert
+    //    the second sort permutation to produce an inverse ID mapping. If
+    //    counts are required, also take the adjacent difference between
+    //    segment_ends indices to produce counts.
+    //      output = input[sorted_unique_input_inds]
+    //      inv_sorted_unique_perm[sorted_unique_perm[i]] = i
+    //      counts = adjacent_difference(segment_ends)
+    // 6) Look up unique IDs via the inverse ID mapping and scatter them using
+    //    the original sort permutation to produce the indices output.
+    //      idx[sorted_input_inds] =
+    //          inv_sorted_unique_perm[sorted_input_unique_ids]
```
Would it be too cumbersome to interleaved a small but non-trivial running example with the algorithm?  That would make the algorithm much easier to understand at a glance.

benbarsdell replied:
Done.

sanjoy commented on tensorflow/python/kernel_tests/unique_op_test.py:
```diff
@@ -152,6 +152,24 @@ def testShapeInferenceV2(self):
+    self.assertEqual(len(x), len(tf_idx))
+    self.assertEqual(len(tf_y), len(np.unique(x)))
+
+  def testOrderedByAppearance(self):
+    x = np.array([3, 5, 3, 4, 1, 4, 9, 8, 6, 3, 5, 7, 8, 8, 4, 6, 4, 2, 5, 6])
+    true_y = np.array([3, 5, 4, 1, 9, 8, 6, 7, 2])
+    true_idx = np.array(
+        [0, 1, 0, 2, 3, 2, 4, 5, 6, 0, 1, 7, 5, 5, 2, 6, 2, 8, 1, 6])
+    y, idx = array_ops.unique(x)
+    tf_y, tf_idx = self.evaluate([y, idx])
+    self.assertAllEqual(tf_y, true_y)
+    self.assertAllEqual(tf_idx, true_idx)
```
I suspect these are exercising the existing CPU kernel for `Unique` that works on integers.  If that's true, can you please use floats?

benbarsdell replied:
The type registrations are identical for CPU and GPU, and the tests are running on the GPU for me.

benbarsdell committed:
Apply minor suggestions from PR 47082

- Standardize inline parameter comment syntax.
- Remove begin_bit parameter and rename end_bit to num_bits.
- Replace NULL with nullptr.

benbarsdell committed:
Add running example to algo summary in UniqueOpGPU

benbarsdell committed:
Add gpu_prim_helpers_test with some simple tests

- Adds tests for GpuRadixSort and GpuInclusivePrefixSum.

benbarsdell commented on tensorflow/core/kernels/gpu_prim_helpers.h:
```diff
@@ -0,0 +1,143 @@
+// If keys_out is not required, it can be set to nullptr.
+// If indices_in is nullptr, the range of input indices [0, size) will be used.
+template <typename Tkey, typename Tindex>
+Status GpuRadixSort(OpKernelContext* context, int size, const Tkey* keys_in,
```
The context is required to do the temporary allocations, which is a large part of the benefit of the helper functions. They're intended to be helpers specifically for implementing op kernels (they reside in the same directory); the underlying functions can still be used directly if needed in other parts of the code.

benbarsdell commented on tensorflow/core/kernels/BUILD:
```diff
@@ -501,6 +501,14 @@ cc_library(
     ]),
 )
 
+cc_library(
+    name = "gpu_prim_helpers",
+    hdrs = ["gpu_prim_helpers.h"],
+    deps = if_cuda_or_rocm([
+        ":gpu_prim_hdrs",
+    ]),
+)
```
Done.

benbarsdell commented on tensorflow/core/kernels/gpu_prim_helpers.h:
```diff
@@ -0,0 +1,143 @@
+Status GpuRadixSort(OpKernelContext* context, int size, const Tkey* keys_in,
+                    Tkey* keys_out,            // Optional
+                    const Tindex* indices_in,  // Optional
+                    Tindex* indices_out, int begin_bit = 0,
```
Removed (also changed end_bit to num_bits).

benbarsdell commented on tensorflow/core/kernels/gpu_prim_helpers.h:
```diff
@@ -0,0 +1,143 @@
+  size_t temp_storage_bytes = 0;
+  const auto& cu_stream = GetGpuStream(context);
+  auto err = gpuprim::DeviceRadixSort::SortPairs(
+      NULL, temp_storage_bytes, keys_in, keys_out, indices_in, indices_out,
```
Done.

benbarsdell commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,443 @@
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          (GpuRadixSort(context, uniq_size,
+                        /*keys_in = */ unique_input_inds_ptr,
```
Done.

sanjoy commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,459 @@
+    const T* input_ptr = input.flat<T>().data();
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        (GpuRadixSort(context, input_size, /*keys_in=*/input_ptr,
```
Extra parens not needed right?

benbarsdell replied:
Removed.

sanjoy commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,459 @@
+    // Copy the last element of sorted_input_unique_ids back to the host to
+    // obtain uniq_size.
+    ScratchSpace<TIndex> last_idx_host(context, 1, /*on_host=*/true);
+    if (input_size > 0) {
```
Why not early exit if `input_size` is `0`?  (Please add a comment.)

benbarsdell replied:
Added early exit instead.

sanjoy reviewed (commented):
Only minor comments.

benbarsdell committed:
Remove unneeded parens in UniqueOpGPU

benbarsdell committed:
Add early exit to UniqueOpGPU for empty input

sanjoy commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,468 @@
+      // Sort by input index so that output is in order of appearance.
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          (GpuRadixSort(context, uniq_size,
```
Unnecessary parens

sanjoy commented on tensorflow/core/kernels/unique_op_gpu.cu.cc:
```diff
@@ -0,0 +1,468 @@
+                                     const TIndex* sorted_input_unique_ids,
+                                     TIndex* unique_input_inds,
+                                     TIndex* segment_ends) {
+  if (input_size == 0) return Status::OK();
```
This can be a `CHECK` right now that you have an early exit for empty input?  Same elsewhere.

benbarsdell committed:
Remove parens and use CHECK macros in UniqueOpGPU

benbarsdell committed:
Remove missing include from gpu_prim_helpers_test

benbarsdell committed:
Add GPU build guard around gpu_prim_helpers_test

benbarsdell committed:
Attempt to fix BUILD for gpu_prim_helpers_test

benbarsdell commented:
I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to https://github.com/tensorflow/tensorflow/pull/46172). Is this important?

[1] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops_impl.h#L277-L280

sanjoy commented:
Hi,

I'm getting failures when I run `tensorflow/python/kernel_tests:unique_op_test_gpu` internally:

```
UniqueWithCountsTest.testInt32
third_party/tensorflow/python/kernel_tests/unique_op_test.py", line 183, in testInt32
    self.assertEqual(len(tf_y), len(np.unique(x)))
AssertionError: 2 != 8
```

(I had to change the `tf_py_test` to a `cuda_py_test` to run the test internally.  I don't think that's necessary in OSS, but you might want to do the same to avoid any confounding issues.)

benbarsdell commented:
Hm, what GPU is it running on when it fails? Also is it just that test or many?

sanjoy commented:
> I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to #46172). Is this important?

Yes, that's important, sorry for not catching that.  (The continuation could be run on another thread.)

sanjoy commented:
> > I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to #46172). Is this important?
> 
> Yes, that's important, sorry for not catching that. (The continuation could be run on another thread.)

CC @kaixih I believe your PR also needs this fix, can you please send a new PR addressing this?  Thanks!

sanjoy commented:
> Hm, what GPU is it running on when it fails? Also is it just that test or many?

It was running on P100, and TF was compiled with clang.

All of these tests fail:

```
UniqueTest.testInt32
UniqueTest.testInt32OutIdxInt64
UniqueTest.testOrderedByAppearance
UniqueWithCountsTest.testInt32
UniqueWithCountsTest.testInt32OutIdxInt64
UniqueWithCountsTest.testOrderedByAppearance
```

And the failure mode is the same -- the output has only two elements.

E.g.

```
third_party/tensorflow/python/kernel_tests/unique_op_test.py", line 49, in testInt32OutIdxInt64
    self.assertEqual(len(tf_y), len(np.unique(x)))
AssertionError: 2 != 8
```

benbarsdell committed:
Add ScopedActivateExecutorContext in UniqueOpGPU

benbarsdell commented:
I managed to build using clang-12, sm_60, and CUDA 11.0u1, but still haven't been able to repro the failure.

I don't suppose changing the `bool` on this line to `TIndex` helps? https://github.com/tensorflow/tensorflow/pull/47082/files#diff-39b40cf87839d63bdeeffbef44effbbd5ff5f3549006091b32dbdfa27f8b8ca5R294

sanjoy commented:
> I managed to build using clang-12, sm_60, and CUDA 11.0u1, but still haven't been able to repro the failure.
> 
> I don't suppose changing the `bool` on this line to `TIndex` helps? https://github.com/tensorflow/tensorflow/pull/47082/files#diff-39b40cf87839d63bdeeffbef44effbbd5ff5f3549006091b32dbdfa27f8b8ca5R294

This link does not point to a line.

benbarsdell commented:
I think you need to expand the contracted diff of `unique_op_gpu.cu.cc` to see the line.

It's line 294 here:
```
    // Create a fancy input iterator to indicate segment boundaries.
    gpuprim::TransformInputIterator<bool, SegmentIndicatorFunctor<T, TIndex>,
                                    gpuprim::CountingInputIterator<TIndex>>
        segment_indicator_iter(0, {sorted_input_ptr});
```
changing to this:
```
    gpuprim::TransformInputIterator<TIndex, SegmentIndicatorFunctor<T, TIndex>,
```

sanjoy commented:
> I think you need to expand the contracted diff of `unique_op_gpu.cu.cc` to see the line.
> 
> It's line 294 here:
> 
> ```
>     // Create a fancy input iterator to indicate segment boundaries.
>     gpuprim::TransformInputIterator<bool, SegmentIndicatorFunctor<T, TIndex>,
>                                     gpuprim::CountingInputIterator<TIndex>>
>         segment_indicator_iter(0, {sorted_input_ptr});
> ```
> 
> changing to this:
> 
> ```
>     gpuprim::TransformInputIterator<TIndex, SegmentIndicatorFunctor<T, TIndex>,
> ```

Wow, that seems to work.  Trying to check in now.

benbarsdell committed:
Fix cause of internal test failure in UniqueOp GPU

- The tests only failed internally at Google, not in public CI.
  It appears that the CUB computation was being done using the
  input type (bool) instead of the output type. Possibly related
  to CUB version?

benbarsdell commented:
I see this was merged, but then reverted in [this nondescript commit](https://github.com/tensorflow/tensorflow/commit/b6a25c7641ac37426b1dadd21b74f814c36f1738)?

I've pushed the `bool` fix to this branch so that it's complete; but now it says there is a merge conflict? I'm a bit confused.

sanjoy commented:
Ben, this was auto-reverted because `//tensorflow/python/debug:analyzer_cli_test_gpu` failed.  The failure seems trivial, I'll try to reland with a fix.

sanjoy commented:
Went in as https://github.com/tensorflow/tensorflow/commit/02585ace8ca787613637713aebfd581a6dc2a12a.
